File: leetcode_75/Binary_Tree_BFS/cpp/199.cpp
#include <vector>
#include <deque>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        if (root == nullptr)
        {
            return {};
        }

        vector<int> result;
        deque<TreeNode *> q = {root};
        TreeNode *node = nullptr;
        while (!q.empty())
        {
            int q_size = q.size();
            for (int i = 0; i < q_size; ++i)
            {
                node = q.front();
                q.pop_front();
                if (node->left)
                {
                    q.push_back(node->left);
                }
                if (node->right)
                {
                    q.push_back(node->right);
                }
            }
            result.push_back(node->val);
        }

        return result;
    }
};

File: Top_Interview_150/Binary_Tree_BFS/python/199.py
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        result = list()

        q = deque([])
        if root:
            q.append(root)

        while q:
            result.append(q[-1].val)

            level_length = len(q)
            for _ in range(level_length):
                node = q.popleft()

                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)

        return result
